Feature/mentor volunteer support

Ciuffi opened the pull request:
I refactored the hacker class to an ApplicantInfo class which allows us to reuse it for volunteer, hacker, mentor. I also added a static map that is setup to match applicantTypes from the database with their collection names.

applicantType defaults to hacker if no type is given.

Ciuffi committed:
changed permissions

Ciuffi committed:
Fix multidex issue and null pointer exception

Ciuffi committed:
Refactor Hacker to ApplicantInfo, add ability to write multiple records

Ciuffi committed:
Simplify collection picking and allow reading from other collections

chamkank commented on app/src/main/java/io/nwhacks/nfc/ApplicantInfo.java:
```diff
@@ -0,0 +1,22 @@
+    public Map<String, Integer> events;
+    public static Map<String, String> applicantMap;
+
+    static {
```
I get the an error when trying to read/write a tag, here's the root error:

```
Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'java.lang.Object java.util.Map.put(java.lang.Object, java.lang.Object)' on a null object reference
        at io.nwhacks.nfc.ApplicantInfo.<clinit>(ApplicantInfo.java:17)
```

I'm not sure if it's working on your end, but it looks like you have to instantiate the `applicantMap` Map first. Seems like there are a couple ways to do this, here's a sensible solution I found: https://stackoverflow.com/a/509016

Ciuffi replied:
I fixed this by using an ImutableMap to create the map statically.

let me know if it works for you.

chamkank replied:
It's working!

chamkank commented on app/src/main/java/io/nwhacks/nfc/NFCManager.java:
```diff
@@ -83,22 +95,25 @@ public Tag getTagFromIntent(Intent intent){
-            return new NdefMessage(new NdefRecord[]{record});
-        } catch (Exception e){
-            return null;
+    NdefMessage createTextMessage(String[] messages) {
```
Nice refactor 💯

chamkank commented on app/src/main/java/io/nwhacks/nfc/ReadFragment.java:
```diff
@@ -122,18 +122,24 @@ public void tagDiscovered(NFCManager mgr, Intent intent) {
+        if (records.get(1) != null){
+            applicantCollection = ApplicantInfo.applicantMap.get(records.get(1));
+        }else{
+            applicantCollection = "hacker_short_info";
```
Instead of defaulting to `hacker`, we probably want to know when/if this issue occurs because it means that the tag has been tampered with, or the writing of `applicantType` to the tag is not working. Can we toast some sort of error message  like `"Unexpected tag contents, please rewrite applicant's tag."` when this happens and return from the method?

Ciuffi replied:
Sure, sounds good!

chamkank reviewed (commented):
Left some comments, also I think you forgot to refactor `onEventJoin` which is called in the `tagDiscovered` method. Right now it's hard-coded to write to the hacker collection only. https://github.com/nwplus/nfc/blob/a0ffcf8b495fcee56fb8d4efcf63992aa3139272/app/src/main/java/io/nwhacks/nfc/ReadFragment.java#L182-L193

Ciuffi committed:
Properly initialize map and check whether a applicanttype tag exists

Ciuffi committed:
add toast error messages for missing applicantType on RFID chip

chamkank commented on app/src/main/java/io/nwhacks/nfc/WriteFragment.java:
```diff
@@ -101,6 +104,14 @@ public void onEvent(@Nullable DocumentSnapshot documentSnapshot, @Nullable Fireb
+                        writeType.setText(di.writeApplicantType);
+                        setColor(DEFAULT_COLOR);
+                    }else{
+                        MainActivity.toast(getContext(), "No ApplicantType for this applicant. Please check that you've selected an applicant.");
```
Great, this covers the missing record case. Can we deal with the case that the second record does exist, but doesn't contain a valid `applicantType` (i.e. tag was tampered with)? You can either detach this else case and make it execute if `applicantCollection == null`, or write a separate if statement after this conditional block.

Ciuffi replied:
I decided to change the larger if statement checking `records.size() > 0` to `applicantType != null`

I've also moved the if statement to the end of this larger if statement.

Ciuffi committed:
cleanup applicantType checking

chamkank commented on app/src/main/java/io/nwhacks/nfc/ReadFragment.java:
```diff
@@ -172,6 +170,9 @@ public void onFailure(@NonNull Exception e) {
                         }
                     });
+        }else {
+            MainActivity.toast(getContext(), "No ApplicantType for this applicant. Please rewrite id.");
```
Minor: "Please rewrite tag." might be more accurate

chamkank replied:
Besides that, it's just `onEventJoin` that needs to be refactored

Ciuffi replied:
fixed!

Ciuffi committed:
Move ApplicantType textview to read instead of write

There isn't really a point in letting the writer know whether they're writing mentor or volunteer because they'll have selected the form on their computer. The reader however might find this information useful.

Ciuffi committed:
id -> tag in toast warning

Ciuffi committed:
ReAdd applicantType to write mode

Ciuffi committed:
fix write_fragment formatting after adding applicantType

Ciuffi committed:
Fixed issue where volunteer and mentor collections weren't being checked

Ciuffi committed:
More EventJoin fixes

Ciuffi commented:
@chamkank I think this PR is essentially ready to go.

There seems to be a pretty big problem with the mentor page though. I can't seem to select a specific mentor to update the app with. Hacker and volunteer forms work fine.

chamkank reviewed (approved):
Just tested it, LGTM! 💯 

As for the mentor bug, it was in part due to the issue mentioned in https://github.com/nwplus/nwhacks2019/pull/246, but now there's an issue with long info being null (since there is no `mentor_long_info`). I'm about to push out a fix for that, thanks for letting me know.
